File: rascaline/tests/data/soap-power-spectrum.py
import ase
import numpy as np

from rascaline import SoapPowerSpectrum
from save_data import save_calculator_input, save_numpy_array


frame = ase.Atoms(
    "CHOC4H2",
    positions=[
        [3.88997, 5.11396, 1.9859],
        [1.60538, 5.74085, 3.48071],
        [5.15178, 5.59335, 5.55114],
        [2.22548, 2.03678, 4.16896],
        [2.33853, 2.79487, 2.3533],
        [3.54073, 3.59016, 2.34664],
        [1.34344, 2.94555, 2.4665],
        [1.74165, 3.03466, 0.921584],
        [0.474942, 3.34246, 2.73754],
    ],
    pbc=[False, False, False],
)

hyperparameters = {
    "cutoff": {
        "radius": 5.5,
        "smoothing": {"type": "ShiftedCosine", "width": 0.5},
    },
    "density": {
        "type": "Gaussian",
        "width": 0.3,
    },
    "basis": {
        "type": "TensorProduct",
        "max_angular": 8,
        "radial": {"max_radial": 7, "type": "Gto"},
        "spline_accuracy": None,
    },
}

calculator = SoapPowerSpectrum(**hyperparameters)
descriptor = calculator.compute(frame, use_native_system=True)
descriptor = descriptor.keys_to_samples("center_type")
descriptor = descriptor.keys_to_properties(["neighbor_1_type", "neighbor_2_type"])

save_calculator_input("soap-power-spectrum-values", frame, hyperparameters)
save_numpy_array("soap-power-spectrum-values", descriptor.block().values)

# Use less values for gradients to keep the file size low
hyperparameters["basis"]["radial"]["max_radial"] = 2
hyperparameters["basis"]["max_angular"] = 4

frame.cell = [6.0, 6.0, 6.0]
frame.pbc = [True, True, True]

calculator = SoapPowerSpectrum(**hyperparameters)
descriptor = calculator.compute(
    frame,
    use_native_system=True,
    gradients=["positions", "strain", "cell"],
)
descriptor = descriptor.keys_to_samples("center_type")
descriptor = descriptor.keys_to_properties(["neighbor_1_type", "neighbor_2_type"])


def sum_gradient(descriptor):
    """compute the gradient w.r.t. each atom of the sum of all rows"""
    gradient = descriptor.block().gradient("positions")

    result = np.zeros((len(frame), 3, len(gradient.properties)))
    for sample, gradient in zip(gradient.samples, gradient.values):
        result[sample["atom"]] += gradient

    return result


save_calculator_input("soap-power-spectrum-gradients", frame, hyperparameters)
save_numpy_array("soap-power-spectrum-positions-gradient", sum_gradient(descriptor))

strain_gradient = descriptor.block().gradient("strain").values
save_numpy_array("soap-power-spectrum-strain-gradient", strain_gradient)

cell_gradient = descriptor.block().gradient("cell").values
save_numpy_array("soap-power-spectrum-cell-gradient", cell_gradient)

File: featomic/benches/soap-power-spectrum.rs
#![allow(clippy::needless_return)]
use featomic::{Calculator, System, CalculationOptions};
use chemfiles::{Frame, Trajectory};

use criterion::{BenchmarkGroup, Criterion, measurement::WallTime, SamplingMode};
use criterion::{criterion_group, criterion_main};


fn read_systems_from_file(path: &str) -> Vec<Box<dyn System>> {
    let mut trajectory = Trajectory::open(path, 'r').expect("could not open the trajectory");
    let mut frame = Frame::new();
    let mut systems = Vec::new();
    for step in 0..trajectory.nsteps() {
        trajectory.read_step(step, &mut frame).expect("failed to read single frame");
        systems.push((&frame).into());
    }

    systems
}

fn run_soap_power_spectrum(
    mut group: BenchmarkGroup<WallTime>,
    path: &str,
    gradients: bool,
    test_mode: bool,
) {
    let mut systems = read_systems_from_file(&format!("benches/data/{}", path));

    if test_mode {
        // Reduce the time/RAM required to test the benchmarks code.
        // Without this, the process gets killed in github actions CI
        systems.truncate(1);
    }

    let cutoff = 4.0;
    let mut n_centers = 0;
    for system in &mut systems {
        n_centers += system.size().unwrap();
        system.compute_neighbors(cutoff).unwrap();
    }

    for &max_basis in &[1, 7, 14] {
        // keep the memory requirements under control
        if max_basis == 14 {
            systems.truncate(10);
            n_centers = 0;
            for system in &mut systems {
                n_centers += system.size().unwrap();
            }
        }

        let parameters = format!(r#"{{
            "cutoff": {{
                "radius": {cutoff},
                "smoothing": {{
                    "type": "ShiftedCosine",
                    "width": 0.5
                }}
            }},
            "density": {{
                "type": "Gaussian",
                "width": 0.3
            }},
            "basis": {{
                "type": "TensorProduct",
                "max_angular": {max_basis},
                "radial": {{
                    "type": "Gto",
                    "max_radial": {max_basis}
                }}
            }}
        }}"#);

        let mut calculator = Calculator::new("soap_power_spectrum", parameters).unwrap();

        group.bench_function(format!("max_radial = max_angular = {}", max_basis), |b| b.iter_custom(|repeat| {
            let start = std::time::Instant::now();

            let options = CalculationOptions {
                gradients: if gradients { &["positions"] } else { &[] },
                ..Default::default()
            };

            for _ in 0..repeat {
                calculator.compute(&mut systems, options).unwrap();
            }
            start.elapsed() / n_centers as u32
        }));
    }
}

fn soap_power_spectrum(c: &mut Criterion) {
    let test_mode = std::env::args().any(|arg| arg == "--test");

    let mut group = c.benchmark_group("SOAP power spectrum (per atom)/Bulk Silicon");
    group.noise_threshold(0.05);
    group.sampling_mode(SamplingMode::Flat);
    group.sample_size(10);

    run_soap_power_spectrum(group, "silicon_bulk.xyz", false, test_mode);

    let mut group = c.benchmark_group("SOAP power spectrum (per atom) with gradients/Bulk Silicon");
    group.noise_threshold(0.05);
    group.sampling_mode(SamplingMode::Flat);
    group.sample_size(10);

    run_soap_power_spectrum(group, "silicon_bulk.xyz", true, test_mode);

    let mut group = c.benchmark_group("SOAP power spectrum (per atom)/Molecular crystals");
    group.noise_threshold(0.05);
    group.sampling_mode(SamplingMode::Flat);
    group.sample_size(10);

    run_soap_power_spectrum(group, "molecular_crystals.xyz", false, test_mode);

    let mut group = c.benchmark_group("SOAP power spectrum (per atom) with gradients/Molecular crystals");
    group.noise_threshold(0.05);
    group.sampling_mode(SamplingMode::Flat);
    group.sample_size(10);

    run_soap_power_spectrum(group, "molecular_crystals.xyz", true, test_mode);
}


criterion_group!(all, soap_power_spectrum);
criterion_main!(all);
